### PR TITLE
Propagate defines into CFLAGS and CXXFLAGS passed to the framework

### PR DESCRIPTION
**Pull request by jsharpe**
(Hopefully) Addresses issue highlighted in #301.

I suspect there are similar issues in propagating includes from deps `CcInfo.compilation_context`.

### CODE REVIEW

**Comment by UebelAndre** [comment]
Can we also incorporate the one of the simpler examples in that issue into this PR? I think it'd be valuable.

**Comment by jsharpe** [comment]
> Can we also incorporate the one of the simpler examples in that issue into this PR? I think it'd be valuable.

The examples as is in the issue is incomplete (missing at least openssl from the deps for CURL), also a compilation only example wouldn't necessarily catch a regression - we'd need to explicitly check that curl was in fact being built with the correct defines set.

**Comment by UebelAndre** [comment]
Wouldn't something like the following work?
```
                    +---------------+
  +->  lib_b  +-->  | lib_a         |
  |                 +---------------+
  +                 |               |
-DFOO               | #ifndef FOO   |
                    | #error NoFoo  |
                    | #endif        |
                    |               |
                    +---------------+

```

**Comment by wrowe** [comment]
This appears to resolve #301 after inspecting the .obj.params intermediate input files (still building, so can't entirely confirm.)

The question I've had, for some time, is how to use select at line 305 much as I did at line 310 below. Obviously, select can't be used for dictionaries, only for lists. My thought is that adding copts accepting a list, same as linkopts, gets around this headache. Any other thoughts? 
https://github.com/wrowe/envoy/blob/9b46e213ddb8a32d838732da045f71a4af40653c/bazel/foreign_cc/BUILD#L305

**Comment by wrowe** [comment]
So, might this also address the missing -fPIC flag? Observed by @lizan in https://github.com/bazelbuild/rules_foreign_cc/issues/421

**Comment by jsharpe** [comment]
> So, might this also address the missing -fPIC flag? Observed by @lizan in #421

Unfortunately not - I think that #421 a separate issue that needs some investigation - we might need a `config_setting` to lookup the value of `--force_pic` and manually enable the `pic` feature from the toolchain, although this might be toolchain specific if you're not using the default toolchain.

**Comment by jsharpe** [comment]
@UebelAndre PTAL - I've added a build test along the lines you suggested and verified it fails if I remove this patch.

File: .bazelci/config.yaml [hunk 1]
```diff
@@ -27,6 +27,7 @@ tasks:
     working_directory: examples
     build_targets:
       - "//cmake_android:app"
+      - "//cmake_defines:all"
```
Comment (UebelAndre) [inline .bazelci/config.yaml:30]: Can't wait to upgrade the min version to 4.0 so we can just build `//...`

**Review by UebelAndre — APPROVED** [review]
I didn't know that `examples/cmake_defines/lib_b/lib_b.cpp` could be empty. I probably wrote more boiler plate than I had to in my time 😆 

Looks good! Thanks for adding the tests 🙏

**Pull request merged**